wasm: Add how to handle WASM documentation (#942)
fluent/fluent-bit-docs: development/wasm-filter-plugins.md

# WASM Filter Plugins

> Fluent Bit currently supports integration of wasm plugins built as wasm/wasi objects for input and filter plugins only. The interface for the WASM filter plugins is currently under development but is functional.

Reviewer: @cosmo0920 can you introduce WASM a little first maybe?

Also a sentence per line is easier to review.

Author: I added in https://github.com/fluent/fluent-bit-docs/pull/942/commits/95ee5c271d79d5ffddae09578bf0191945c453e9.

## Prerequisites

> ### For Build fluent-bit

Reviewer: ```suggestion
### Building Fluent Bit
```

> For only execution of WASM input and filter plugins, there is no additional requirement.

Reviewer: ```suggestion
There are no additional requirements to execute WASM plugins.
```
No need to repeat filter and input here as when/if it changes we will have to remove it.

#### Building flb-wamrc (optional)

> For `flb-wamrc` support, users have to install llvm infrastructure, e.g:

Reviewer: What is this? Can you introduce or link?

Author: I added a description for `flb-wamrc` as:

> `flb-wamrc` is just `flb-` prefixed AOT (Ahead Of Time) compiler that is provided from [wasm-micro-runtime](https://github.com/bytecodealliance/wasm-micro-runtime).

> ```text

Reviewer: ```suggestion
\```shell
```

# apt install -y llvm
```

### For Build WASM programs

Currently, fluent-bit supports the following WASM toolchains:

* Rust on `wasm32-unknown-unknown`.
  * rustc 1.62.1 (e092d0b6b 2022-07-16) or later
* [TinyGo](https://github.com/tinygo-org/tinygo) on wasm32-wasi
  * v0.24.0 or later
* [WASI SDK](https://github.com/WebAssembly/wasi-sdk) 13 or later.

## Getting Started

> Compile Fluent Bit with WASM support, e.g:

Reviewer: Can we just reference the other docs on building?

Author: AFAIK, I'd never seen the complete documentation to build fluent-bit in other fluent-bit doc.

Reviewer 2: we should add it under this list https://docs.fluentbit.io/manual/installation/sources/build-and-install#build-options

Author: Added in https://github.com/fluent/fluent-bit-docs/pull/942/commits/69ef376bf0c072db422c237e4ba741ccab971de8

```text
$ cd build/
$ cmake .. [-DFLB_WAMRC=On]
$ make
```

To support AOT compiled WASM execution as filter plugins, users have to built fluent-bit with `-DFLB_WAMRC=On`.

Once compiled, we can see new plugins in which handles _wasm_, e.g:

```text
$ bin/fluent-bit -h
Usage: fluent-bit [OPTION]
Inputs
  # ... other input plugin stuffs
  exec_wasi               Exec WASI Input

Filters
  # ... other filter plugin stuffs
  wasm                    WASM program filter
```

## Build a WASM Filter for Filter Plugin

> Currently, fluent-bit's WASM filter assumes C ABI that is also known as `wasm32-unknown-unknown` on Rust target and `wasm32-wasi` on TinyGo target.

Reviewer: Need to be consistent in the docs with "Fluent Bit" and not mix it up.

```text
$ rustup target add wasm32-unknown-unknown
```

### Requirements of WASM programs

WASM filter plugins execute the function that has the following signagure.

For C:

```text
// We can use an arbitrary function name for filter operations w/ WASM.
char* c_filter(char*, int, uint32_t, uint32_t, char*, int);
```

For Go(TinyGo):

```text
//export go_filter
// And this function should be placed in the main package.
func go_filter(tag *uint8, tag_len uint, time_sec uint, time_nsec uint, record *uint8, record_len uint) *uint8
```

For Rust:

```text
// #[no_mangle] attribute is needed for preventing mangles and align C ABI.
// Also we can use an arbitrary function name for filter operations w/ WASM.
#[no_mangle]
pub extern “C” fn rust_filter(tag: *const c_char, tag_len: u32, time_sec: u32, time_nsec: u32, record: *const c_char, record_len: u32)
```

Note that `//export XXX` on TinyGo and `#[no_mangle]` attributes on Rust are required. This is because TinyGo and Rust will mangle their function names if they are not specified.

> Once built, a WASM program will be available. Then, that built program can be executed with the following fluent-bit configuration:

Reviewer: Again, more "Fluent Bit" changes.

Can you also show the output of this? Maybe a full WASM code example or a link to the one we have then the output?

Author: Sure, I'll do it later.

```text
[SERVICE]
    Flush        1

>     Daemon       Off

Reviewer: All this service config seems unnecessary as it is the defaults. Plus we turn off http server but then pass it a port and interface.

    Log_Level    info
    HTTP_Server  Off
    HTTP_Listen  0.0.0.0
    HTTP_Port    2020

[INPUT]
    Name dummy
    Tag dummy.local

[FILTER]
    Name wasm
    Tag  dummy.*
    WASM_Path /path/to/built_filter.wasm
    Function_Name super_awesome_filter
    accessible_paths .,/path/to/fluent-bit

[OUTPUT]
    Name  stdout
    Match *
```

### Optimize execution of WASM programs

> To optimize WASM program execution, users can use `flb-wamrc` for this purpose.

Reviewer: ```suggestion
To optimize WASM program execution, there is the option of using `flb-wamrc`.
```

This tool will be built when `-DFLB_WAMRC=On` cmake option is specififed and llvm infrastructure is installed on the building box.

```text
$ flb-wamrc -o /path/to/built_wasm.aot /path/to/built_wasm.wasm
```

> For more optimizing the specific CPU such as Cortex-A57 series, e.g:

Reviewer: ```suggestion
For further optimizations to the specific CPU such as Cortex-A57 series, e.g:
```

```text
$ flb-wamrc --size-level=3 --target=aarch64v8 --cpu=cortex-a57 -o /path/to/built_wasm.aot /path/to/built_wasm.wasm
```

> Then, when AOT compiling is succeeded:

Reviewer: What is AOT?
